src/app/qgsdockablewidgethelper.cpp

```diff
@@ -0,0 +1,258 @@
+/***************************************************************************
+  qgsdockablewidgethelper.cpp
+  --------------------------------------
+  Date                 : January 2022
+  Copyright            : (C) 2022 by Belgacem Nedjima
+  Email                : belgacem dot nedjima at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsdockablewidgethelper.h"
+
+#include "qgsdockwidget.h"
+#include "qgsapplication.h"
+
+#include <QLayout>
+
+QgsDockableWidgetHelper::QgsDockableWidgetHelper( bool isDocked, const QString &windowTitle, QWidget *widget, QMainWindow *ownerWindow )
+  : QObject( nullptr )
+  , mWidget( widget )
+  , mDialogGeometry( 0, 0, 200, 200 )
+  , mDockGeometry( QRect() )
+  , mIsDockFloating( true )
+  , mDockArea( Qt::RightDockWidgetArea )
+  , mWindowTitle( windowTitle )
+  , mOwnerWindow( ownerWindow )
+{
+  toggleDockMode( isDocked );
+  mToggleButton.setIcon( QgsApplication::getThemeIcon( QStringLiteral( "mDockify.svg" ) ) );
+  mToggleButton.setToolTip( tr( "Dock 3D Map View" ) );
+  mToggleButton.setCheckable( true );
+  mToggleButton.setChecked( isDocked );
+  mToggleButton.setEnabled( true );
```

Comment:
I think rather than having the button like this `mToggleButton`, it will be better to have a function `QToolButton* createToolButton()` which would create it on the fly - for two reasons - 1. client class would be able to use multiple buttons if ever needed, 2. there will be no doubt how the ownership of the object works (now it is owned by the helper class, but also it is somewhere in qobject parent-child hierarchy, which is a bit confusing)